compression_algorithms/huffmann/huffmann.go

```diff
@@ -0,0 +1,275 @@
+package main
+
+import (
+    "container/heap"
+    "fmt"
+    "io/ioutil"
+    "os"
+    "sort"
+    "bufio"
+    "strings"
+)
+
+type Node struct {
+    char   rune
+    freq   int
+    left   *Node
+    right  *Node
+    isLeaf bool
+}
+
+type PriorityQueue []*Node
+
+func (pq PriorityQueue) Len() int { return len(pq) }
+func (pq PriorityQueue) Less(i, j int) bool {
+    if pq[i].freq == pq[j].freq {
+        return pq[i].char < pq[j].char
+    }
+    return pq[i].freq < pq[j].freq
+}
+func (pq PriorityQueue) Swap(i, j int) {
+    pq[i], pq[j] = pq[j], pq[i]
+}
+func (pq *PriorityQueue) Push(x interface{}) {
+    *pq = append(*pq, x.(*Node))
+}
+func (pq *PriorityQueue) Pop() interface{} {
+    old := *pq
+    n := len(old)
+    x := old[n-1]
+    *pq = old[0 : n-1]
+    return x
+}
+
+func BuildHuffmanTree(freqMap map[rune]int) *Node {
+    // Handle empty input
+    if len(freqMap) == 0 {
+        return nil
+    }
+
+    // Handle single character case
+    if len(freqMap) == 1 {
+        for char, freq := range freqMap {
+            return &Node{
+                char:   char,
+                freq:   freq,
+                isLeaf: true,
+            }
+        }
+    }
+
+    pq := &PriorityQueue{}
+    heap.Init(pq)
+    
+    chars := make([]rune, 0, len(freqMap))
+    for char := range freqMap {
+        chars = append(chars, char)
+    }
+    
+    sort.Slice(chars, func(i, j int) bool {
+        return chars[i] < chars[j]
+    })
```

Comment:
Why is sorting done here?
Won't appending them to heap take care of that?

Reply:
Yes the logic has been corrected